ui/next/app/redux/databaseInfo.ts

```diff
@@ -0,0 +1,374 @@
+export const TABLE_DETAILS_INVALIDATE = "cockroachui/databases/TABLE_DETAILS_INVALIDATE";
+
+export class DatabasesState {
+  data: DatabasesResponseMessage;
+  inFlight = false;
+  valid = false;
+  lastError: Error;
+}
+
+export class DatabaseDetailsState {
+  data: DatabaseDetailsResponseMessage;
+  inFlight = false;
+  valid = false;
+  lastError: Error;
+}
+
+export class TableDetailsState {
+  data: TableDetailsResponseMessage;
+  inFlight = false;
+  valid = false;
+  lastError: Error;
+}
+
+export class AllDatabaseDetailsState {
+  [id: string]: DatabaseDetailsState
+}
+
+export class AllTableDetailsState {
+  [id: string]: DatabaseDetailsState
+}
+
+/**
+ * Represents the current state of a database list query.
+ */
+export class DatabaseInfoState {
+  databases: DatabasesState;
+  databaseDetails: AllDatabaseDetailsState;
+  tableDetails: AllTableDetailsState;
+}
+
+/**
+ * Redux reducer which processes actions related to the database list query.
+ */
+export function databasesReducer(state: DatabasesState = new DatabasesState(), action: Action): DatabasesState {
+  switch (action.type) {
+    case DATABASES_REQUEST:
+      // A request is in progress.
+      state = _.clone(state);
+      state.inFlight = true;
+      return state;
+    case DATABASES_RECEIVE:
+      // The results of a request have been received.
+      let { payload } = action as PayloadAction<cockroach.server.DatabasesResponseMessage>;
+      state = _.clone(state);
+      state.inFlight = false;
+      state.data = payload;
+      state.valid = true;
+      state.lastError = null;
+      return state;
+    case DATABASES_ERROR:
+      // A request failed.
+      let { payload: error } = action as PayloadAction<Error>;
+      state = _.clone(state);
+      state.inFlight = false;
+      state.lastError = error;
+      state.valid = false;
+      return state;
+    case DATABASES_INVALIDATE:
+      state = _.clone(state);
+      state.valid = false;
+      return state;
+    default:
+      return state;
+  }
+}
+
+interface WithID<T> {
+  id: string;
+  data: T;
+}
+
+/* TODO: @mrtracy - you mentioned a generic cached data reducer. This should be
+   added to that as well.
+*/
+
+export function databaseDetailsReducer(state: AllDatabaseDetailsState = new AllDatabaseDetailsState(), action: Action): AllDatabaseDetailsState {
+  switch (action.type) {
+    case DATABASE_DETAILS_REQUEST:
+      let { payload: id } = action as PayloadAction<string>;
+      state = _.clone(state);
+      state[id] = _.clone(state[id]) || new DatabaseDetailsState();
+      state[id].inFlight = true;
+      return state;
+    case DATABASE_DETAILS_RECEIVE:
+      // The results of a request have been received.
+      let { payload } = action as PayloadAction<WithID<DatabaseDetailsResponseMessage>>;
+      state = _.clone(state);
+      state[payload.id] = _.clone(state[payload.id]) || new DatabaseDetailsState();
```

Comment:
What's the benefit of this way over the other? Since essentially the same thing is taking place.